src/managers/MilestoneEventManager.ts

```diff
@@ -0,0 +1,179 @@
+import { checkCodeTimeMetricsMilestonesAchieved, checkDaysMilestones, checkLanguageMilestonesAchieved, compareWithLocalMilestones, getMilestonesByDate } from "../utils/MilestonesUtil";
+import { syncSummary } from "../utils/SummaryUtil";
+import { getItem, isLoggedIn } from "../utils/Util";
+import { isResponseOk, softwareGet, softwarePost } from "./HttpManager";
+
+export class MilestoneEventManager {
+	private static instance: MilestoneEventManager;
+
+	private _checkMilestoneTimeout: any;
+	private _checkingForMilestones: boolean = false;
+
+	static getInstance(): MilestoneEventManager {
+		if (!MilestoneEventManager.instance) {
+			MilestoneEventManager.instance = new MilestoneEventManager();
+		}
+
+		return MilestoneEventManager.instance;
+	}
+
+	private constructor() {
+		// document listener handlers
+		workspace.onDidOpenTextDocument(this._onOpenHandler, this);
+		workspace.onDidCloseTextDocument(this._onCloseHandler, this);
+		workspace.onDidChangeTextDocument(this._onEventHandler, this);
+		// window state changed handler
+		window.onDidChangeWindowState(this._windowStateChanged, this);
+
+		this.checkMilestonesLazily();
+	}
+
+	private _onCloseHandler(textDoc: TextDocument) {
+		// set a timer to check so we don't disrupt typing
+		this.checkMilestonesLazily();
+	}
+
+	private _onOpenHandler(textDoc: TextDocument) {
+		// set a timer to check so we don't disrupt typing
+		this.checkMilestonesLazily();
+	}
+
+	private _onEventHandler(textDocChangeEvent: TextDocumentChangeEvent) {
+		// set a timer to check so we don't disrupt typing
+		this.checkMilestonesLazily();
+	}
+
+	private _windowStateChanged(winState: WindowState) {
+		this.checkMilestonesLazily();
+	}
+
+	private checkMilestonesLazily() {
+		if (this._checkMilestoneTimeout) {
+			// cancel the current one
+			clearTimeout(this._checkMilestoneTimeout);
+			this._checkMilestoneTimeout = null;
+		}
+		this._checkMilestoneTimeout = setTimeout(() => {
+			this.checkForMilestones();
+		}, 1000 * 30);
+	}
+
+	private async checkForMilestones() {
+		if (!isLoggedIn() || this._checkingForMilestones) {
+			return;
+		}
+		this._checkingForMilestones = true;
+		// updates logs with latest metrics and checks for milestones
+
+		// checks to see if there are any new achieved milestones
+		const achievedTimeMilestones = checkCodeTimeMetricsMilestonesAchieved();
+
+		// checks to see if there are any new language milestones achieved
+		const achievedLangMilestones = checkLanguageMilestonesAchieved();
+
+		// checks to see if there are any day milestones achived
+		const achievedDaysMilestones = checkDaysMilestones();
+
+		if (achievedDaysMilestones.length || achievedLangMilestones.length || achievedTimeMilestones.length) {
+			// get the current milestones so if its an update the milestone array
+			// has them all instead of getting replaced by net new milestones
+			const currentMilestones: Array<number> = getMilestonesByDate(new Date().valueOf()) || [];
```

Comment:
This is probably out of scope, but instead of using the `getMilestonesByDate` function, this could probably just get the most recent day of milestones.

Reply:
yeah, i can look at cleaning that up. the date that does go into there is the most recent day but it can be cleaned up to make sure its readable

Reply:
i've updated that function name as its only used to fetch today's local milestones